lib/if_rmap.c

```diff
@@ -174,78 +165,46 @@ static int if_rmap_unset(struct if_rmap_ctx *ctx,
 	return 1;
 }
 
-DEFUN (if_rmap,
-       if_rmap_cmd,
-       "route-map RMAP_NAME <in|out> IFNAME",
-       "Route map set\n"
-       "Route map name\n"
-       "Route map set for input filtering\n"
-       "Route map set for output filtering\n"
-       "Route map interface name\n")
+static int if_route_map_handler(struct vty *vty, const char *no,
+				const char *dir, const char *ifname,
+				const char *route_map)
 {
-	int idx_rmap_name = 1;
-	int idx_in_out = 2;
-	int idx_ifname = 3;
-	enum if_rmap_type type;
-	struct if_rmap_ctx *ctx;
-	const struct lyd_node *dnode;
-	struct rip *rip;
-
-	dnode = yang_dnode_get(running_config->dnode, VTY_CURR_XPATH);
-	rip = nb_running_get_entry(dnode, NULL, true);
-	ctx = rip->if_rmap_ctx;
-
-	if (strncmp(argv[idx_in_out]->text, "in", 1) == 0)
-		type = IF_RMAP_IN;
-	else if (strncmp(argv[idx_in_out]->text, "out", 1) == 0)
-		type = IF_RMAP_OUT;
-	else {
-		vty_out(vty, "route-map direction must be [in|out]\n");
-		return CMD_WARNING_CONFIG_FAILED;
-	}
+	enum nb_operation op = no ? NB_OP_DESTROY : NB_OP_MODIFY;
+	char xpath[XPATH_MAXLEN];
 
-	if_rmap_set(ctx, argv[idx_ifname]->arg,
-		    type, argv[idx_rmap_name]->arg);
+	snprintf(xpath, sizeof(xpath),
+		 "./if-route-maps/if-route-map[interface='%s']/%s-route-map",
+		 ifname, dir);
+	nb_cli_enqueue_change(vty, xpath, op, route_map);
```

Comment:
In the `no` case there should be an additional check to destroy the list element `./if-route-maps/if-route-map[interface='%s']` if both in/out route-maps are deleted. Currently, it is never deleted and becomes a leaked memory.

Reply:
Actually, this is why your `must "in-route-map or out-route-map";` raises a validation error. libyang doesn't delete the container automatically, you have to do it manually.

Reply:
right. thanks for catching this, fixed.